openhands/runtime/impl/runloop/runloop_runtime.py

```diff
@@ -260,42 +183,26 @@ def close(self, rm_all_containers: bool | None = True):
 
     @property
     def vscode_url(self) -> str | None:
-        if self.vscode_enabled and self.devbox and self.devbox.status == 'running':
-            if self._vscode_url is not None:
-                return self._vscode_url
-
-            try:
-                with send_request(
-                    self.session,
-                    'GET',
-                    f'{self.api_url}/vscode/connection_token',
-                    timeout=10,
-                ) as response:
-                    response_json = response.json()
-                    assert isinstance(response_json, dict)
-                    if response_json['token'] is None:
-                        return None
-                    token = response_json['token']
-
-                self._vscode_url = (
-                    self.runloop_api_client.devboxes.create_tunnel(
-                        id=self.devbox.id,
-                        port=self._vscode_port,
-                    ).url
-                    + f'/?tkn={token}&folder={self.config.workspace_mount_path_in_sandbox}'
-                )
+        if (
+            hasattr(self, '_vscode_url') and self._vscode_url is not None
```

Comment:
I'm not not sure why we have `hasattr` here considering we set it in the constructor

Reply:
good point!